Event definition drawer properties

liyiy opened the pull request:
## Changes

Properties stats for the event definition drawer

<img width="1333" alt="Screen Shot 2021-06-14 at 9 50 42 AM" src="https://user-images.githubusercontent.com/25164963/121903274-23d33f00-ccf6-11eb-9d7a-f2f0b6a85d47.png">


## Checklist

- [ ] All querysets/queries filter by Organization, by Team, and by User
- [ ] Django backend tests
- [ ] Jest frontend tests
- [ ] Cypress end-to-end tests
- [ ] Migrations are safe to run at scale (e.g. PostHog Cloud) – present proof if not obvious
- [ ] Frontend/CSS is usable at 320px (iPhone SE) and decent at 360px (most phones)

liyiy committed:
wrap css classnames and allow for property definition filtering

liyiy committed:
event properties section

liyiy committed:
properties stats table and object tag functionality

liyiy committed:
property descriptions and save all descriptions button

liyiy committed:
fix description editing on the table

liyiy committed:
Merge branch 'master' into event-definition-drawer-properties

liyiy committed:
fix test failures

liyiy committed:
Merge branch 'master' into event-definition-drawer-properties

liyiy committed:
debouncing input is weird

liyiy commented:
`/home/runner/work/_temp/b639176d-47a1-475c-b81a-4483679da149.sh: /home/runner/.virtualenvs/.venv/bin/black:` `/home/runner/.virtualenvs/.venv/bin/python: bad interpreter: No such file or directory`

`cannot find module 'ospath'`

Those are the errors I'm getting for my tests, this PR should still be review-able while I try to get those to stop failing

mariusandra commented on frontend/src/scenes/events/definitions/DefinitionDrawer.tsx:
```diff
@@ -1,6 +1,6 @@
 import { useActions, useValues } from 'kea'
```
This file desperately wants to be split into multiple small files. 

Golden rule: one exported component per file.

mariusandra commented on frontend/src/scenes/events/definitions/definitionDrawerLogic.ts:
```diff
@@ -2,45 +2,86 @@ import { kea } from 'kea'
+                },
+            },
+        ],
+        editedDefinitions: [
+            [] as PropertyDefinition[],
+            {
+                setDefinitionUpdateList: (property: PropertyDefinition) => {
+                    const ids = values.editedDefinitions.flatMap((def) => def.id)
+                    if (ids.includes(property.id)) {
+                        return values.editedDefinitions.map((def) => (def === property.id ? property : def))
+                    }
+                    return [...values.editedDefinitions, property]
+                },
+            },
+        ],
```
This should also be a reducer, not a loader

mariusandra commented on frontend/src/scenes/events/definitions/definitionDrawerLogic.ts:
```diff
@@ -2,45 +2,86 @@ import { kea } from 'kea'
+                    )
+                    return propertyDefinitions.results
+                },
+                setPropertyDefinitions: (newProperty) => {
+                    return values.eventProperties.map((prop) => (prop.id === newProperty.id ? newProperty : prop))
+                },
```
This should be a reducer, not a loader

mariusandra commented on frontend/src/scenes/events/definitions/definitionDrawerLogic.ts:
```diff
@@ -132,20 +179,51 @@ export const definitionDrawerLogic = kea<definitionDrawerLogicType<EventOrPropTy
+            actions.setDefinitionUpdateList(prop)
+        },
+        updateAllDescriptions: async () => {
+            actions.setSaveAllLoading(true)
```
This `setSaveAllLoading` action could be removed and the `saveAllLoading` reducer could set itself to `true` when it sees the `updateAllDescriptions` action.

Not sure how to set it back to "false" though. Probably calling the action with `false` is fine in the end.

mariusandra commented on frontend/src/scenes/events/definitions/definitionDrawerLogic.ts:
```diff
@@ -121,6 +167,7 @@ export const definitionDrawerLogic = kea<definitionDrawerLogicType<EventOrPropTy
                 errorToast('Oops! This tag is already set', 'This event already includes the proposed tag.')
                 return
             }
+            actions.setType('event_definitions')
```
I see a lot of `setType` action calls. This feels like a code smell somehow. If it's set in `openDrawer`, why should it ever change? If it's needed to make the right API calls, the `type` should be inside the action payload, not a "magic global".

liyiy replied:
ok! yeah that was what I was thinking about, passing the url as an argument vs creating it as a global, I wasn't sure which was better practice

liyiy replied:
but I leaned towards making it a global because it felt more kea aligned? haha idk, just my thought process

mariusandra replied:
The trouble with globals is that you can never be sure what the value is. You might set it to one type, but some other action elsewhere might reset it again... and when you read it and expect it to be one thing, it might be something else. End result: updating the wrong type of property.

Mostly things will be fine, but if we can get by with less moving parts, we should. Less parts = less code = simpler app = less bugs.

mariusandra reviewed (changes requested):
Hey @liyiy , great work getting it this far. I'm going to shoot various UX and code comments at random :). Some of them are for the UX in general, not just the latest changes from this PR, so feel free to tackle them in whatever order makes sense. UX comments are below, code comments mostly inline. 

1. It makes sense to also pay attention to your logic type files. This one is kind of red and missing several types. I wish kea-typegen would auto-import all such types (coming... soon 😅), but until then we must manually add them to the type.

![image](https://user-images.githubusercontent.com/53387/122044923-5dcb4080-cddd-11eb-89e8-28474629617f.png)

2. It's totally not obvious where you can and can't click and why. At the very least, the cursor should turn into a pointer if you can click, but that's very touch-hostile (users on tablets/phones don't see the cursor), so I think a button at the end of the row is essential for discoverability.

![2021-06-15 14 02 29](https://user-images.githubusercontent.com/53387/122049264-86a20480-cde2-11eb-9fef-f6ed47c5b9f7.gif)

3. Everything scrolls to top when I add a tag:

![2021-06-15 14 04 34](https://user-images.githubusercontent.com/53387/122049405-ab967780-cde2-11eb-83e5-0ddc6ad948d2.gif)

4. The "autosave" is inconsistent

![2021-06-15 14 05 53](https://user-images.githubusercontent.com/53387/122049626-ec8e8c00-cde2-11eb-8ff0-f22f025a64be.gif)

5. This is actually a JSON object. Perhaps it should look like one?

![image](https://user-images.githubusercontent.com/53387/122049689-fdd79880-cde2-11eb-9ffa-730f7897dd8c.png)

6. This "usage" is so tiny I'm not sure why hide it. I'd move it up together with the other stuff ("first seen", etc).

![image](https://user-images.githubusercontent.com/53387/122049913-3bd4bc80-cde3-11eb-912f-bc89b1489c00.png)

7. The "save" button gives no feedback. A toast would be nice. Auto-closing might be better.

![2021-06-15 14 10 31](https://user-images.githubusercontent.com/53387/122050238-91a96480-cde3-11eb-956f-05b662fab933.gif)

8. A property can have properties. Not sure if this makes sense :)

![2021-06-15 14 13 20](https://user-images.githubusercontent.com/53387/122050664-f795ec00-cde3-11eb-9bf8-df65e273dbbb.gif)

liyiy commented:
@mariusandra 

For #2, on my localhost the rows look like this, does this UX feel less confusing?
<img width="954" alt="Screen Shot 2021-06-15 at 8 59 56 AM" src="https://user-images.githubusercontent.com/25164963/122056900-11b9d500-cdb8-11eb-91c5-2b365e078bd2.png">

For #4, that drawer is basically half autosave and half form right now, I should probably lean towards one or the other, probably towards form. What do you think? I wanted to make it more streamlined with the tags and owner dropdown change but got stuck when thinking about autosaving text input.

#6: The usage dropdown was where the chart would have lived, I can definitely move that elsewhere until I get the chart showing

#8: I haven't configured the property drawer properly yet, but yeah there shouldn't be properties for a property, right? Is that what you're saying?

Thanks for the feedback!

liyiy committed:
split up components

liyiy committed:
refactor definition drawer kea file

mariusandra commented:
2. Hmm, that's not what I saw. Must be some toggle / EE / ...? issue when I was testing locally. This looks much better. 

4. Indeed make it into a traditional form that can be saved with "save". Maybe it makes sense to also add another "cancel" button next to "save". In any case, the golden rule here is that the less surprises (even if well meaning), the better the UX.

6. Aha, right :).

8. Yeah, though it's clear this was still TODO :)

liyiy commented:
@mariusandra  Yep! I think most of the concerns are addressed now, just some linting errors I have to fix, but I've refactored the kea file to be all form instead and it's less convoluted now (I think), fixed the state issues, split up the exports,

liyiy committed:
drawer should only open for taxonomy feature users

liyiy committed:
lint errors

liyiy committed:
prettier

liyiy committed:
property definition filter backend test

liyiy committed:
fix test

liyiy committed:
populate test with definitions..

liyiy committed:
clean up

liyiy committed:
clean up property drawer interactions

mariusandra committed:
Merge branch 'master' into event-definition-drawer-properties

mariusandra commented on frontend/src/scenes/events/definitions/EventPropertiesStats.tsx:
```diff
@@ -0,0 +1,98 @@
+        {
+            title: 'Property',
+            key: 'property',
+            render: function renderProperty({ name }: { name: string }) {
```
Reuse types as much as possible. This function could be `renderProperty({ name }: EventOrPropType)`, or maybe even `PropertyDefinition` or something. Same for all the other `renderStuff` below.

mariusandra commented on frontend/src/scenes/events/definitions/definitionDrawerLogic.ts:
```diff
@@ -50,34 +105,52 @@ export const definitionDrawerLogic = kea<definitionDrawerLogicType<EventOrPropTy
-                setDescriptionEditing: (_, { editing }) => editing,
+                setEventPropertyDefinition: (state, { propertyDefinition, id }) => {
+                    const newDefinitions = state.map((p) =>
+                        p.id === id ? Object.assign({}, { ...p, ...propertyDefinition }) : p
```
`Object.assign({}, { ...p, ...propertyDefinition })` does exactly the same as just `{ ...p, ...propertyDefinition }`.

mariusandra commented on frontend/src/scenes/events/definitions/definitionDrawerLogic.ts:
```diff
@@ -100,52 +173,93 @@ export const definitionDrawerLogic = kea<definitionDrawerLogicType<EventOrPropTy
+        ],
+        eventPropertiesDefinitionTags: [
+            (selectors) => [selectors.eventPropertiesDefinitions],
+            (properties: PropertyDefinition[]): string[] => {
```
You normally don't need to specify types in selectors. They are autodetected from the selector inputs... and adding something manually might lead to subtle bugs if something goes out of sync.

mariusandra reviewed (changes requested):
Gave it another look. Few small code comments inline, and a bunch of UX stuff below:

1. Things that can be clicked (== the entire row) should have a `cursor:pointer` property to indicate that this is possible. Otherwise what separates this table row from any other random background element? There's also no arrow on the "properties stats" page at the end of the row.
![2021-06-17 12 25 20](https://user-images.githubusercontent.com/53387/122379437-26889b00-cf67-11eb-8985-9b794bc0d135.gif)

2. It's still annoying that it always scrolls to top after every change:
![2021-06-17 12 27 26](https://user-images.githubusercontent.com/53387/122379752-6e0f2700-cf67-11eb-9563-c02c746ee2df.gif)

3. Even more, if I'm on the 2nd/3rd/etc page, changing anything goes back to the first, no matter if I use the inline editor or the drawer:
![2021-06-17 12 28 37](https://user-images.githubusercontent.com/53387/122379954-a7479700-cf67-11eb-8081-2932c7214b9b.gif)

4. Until we add a graph (unless coming really soon), it still makes sense to move these up
<img width="848" alt="Screenshot 2021-06-17 at 12 30 38" src="https://user-images.githubusercontent.com/53387/122380242-f8f02180-cf67-11eb-8789-7ca05e046b20.png">

5. I can use "ENTER" in one textarea, but not the other
![2021-06-17 12 31 57](https://user-images.githubusercontent.com/53387/122380370-1624f000-cf68-11eb-88a5-465433df772d.gif)

6. Can't change the number of rows on a page
![2021-06-17 12 33 24](https://user-images.githubusercontent.com/53387/122380532-440a3480-cf68-11eb-8d3b-dfd2e551b8f3.gif)

7. Editing event properties --> won't save their data in a way that makes it update in the list
![2021-06-17 12 35 51](https://user-images.githubusercontent.com/53387/122380977-b549e780-cf68-11eb-88c3-5b8a77fe48a9.gif)

8. Same if just editing them on the properties page
![2021-06-17 12 37 35](https://user-images.githubusercontent.com/53387/122381144-df9ba500-cf68-11eb-86d4-d314624b38ec.gif)

liyiy committed:
fix everything but scroll top issue

liyiy committed:
Merge branch 'event-definition-drawer-properties' of https://github.com/PostHog/posthog into event-definition-drawer-properties

liyiy committed:
Merge branch 'master' into event-definition-drawer-properties

liyiy committed:
test scroll issue

liyiy commented:
@mariusandra 
I think I need another pair of eyes on the state issue I'm having with trying to get the table not to scroll back up every time an edit is made, originally I was returning updated definitions in the loader (which is why it kept scrolling) but now when I return it in the reducer, it doesn't trigger a re-render even though the referential equality of the new array of definitions is different

liyiy committed:
table state wip

liyiy committed:
Merge branch 'master' into event-definition-drawer-properties

mariusandra committed:
Merge branch 'master' into event-definition-drawer-properties

mariusandra committed:
update logic after typegen

mariusandra committed:
prevent undefined error

mariusandra committed:
only make the clickable rows clickable, not the entire table

mariusandra committed:
upgrade typegen and fix error with feature flags

mariusandra committed:
move to reducers and rename action

mariusandra committed:
update blacklist

mariusandra committed:
Merge branch 'master' into event-definition-drawer-properties

mariusandra commented:
I spent 2h with this now, but not directly. Found some issues, solved them, found problems that have their root in kea-typegen, fixed those, came back... and ran out of time :).

Currently still broken:
- table doesn't update when editing properties
- properties don't save at all when saving them inside an event (when editing on page 2)
- editing properties saves the event's description inside the property description, no matter what was originally written

I'll give it another look later tonight, however one extra tip:
- I see you used console.log in reducers. That's fine and all, but it's much easier to use the [redux devtools](https://kea.js.org/docs/guide/debugging) to see what actions were fired and what did they change.

liyiy committed:
fix event properties description bug

liyiy commented:
@mariusandra I fixed the property save/edit issue, the issue with the state not making a re-render is still up but since this is behind a feature flag anyway, I'd like to push for this to get merged and to revisit the re-render problem/other bugs in a future PR

mariusandra committed:
update typegen once more

mariusandra committed:
fix editing for events table

mariusandra committed:
fix editing for properties table

mariusandra reviewed (approved):
Alright, let's get this in!

There were [some](https://github.com/PostHog/posthog/pull/4722/commits/aba0aaa8431fe1d1c125d2a39a45e9023c4410d6) [things](https://github.com/PostHog/posthog/pull/4722/commits/a751e47527e911421225b3785b8d5db165abcf55) that were still missing, but now it looks good!

- The entire table used an `useEffect(..., [])` to cache the data for its entire duration
- The inline textareas user `useState` to store the description on load and never changed it
- You couldn't save property descriptions anymore after a "fix", so fixed that as well :)
- I also tagged along a few typegen fixes as well :). We're down to 33 TS errors.

mariusandra committed:
add script to clean typegen

mariusandra committed:
update ts blacklist
